Debugger: Add support for on demand breakpoint expressions

ahelwer opened the pull request:
Getting this out there so we can play around with it and build on the idea, see if it's even viable for the interpreter to use this operator definition, etc.

TODO:
- [ ] Conditionally restrict operator action level depending on breakpoint location
- [x] Implement functionality to find all variables local to the breakpoint scope and add them to the operator definition

Closes #1096

lemmy commented on tlatools/org.lamport.tlatools/src/tla2sany/semantic/ModuleNode.java:
```diff
@@ -163,6 +163,12 @@ public class ModuleNode extends SymbolNode {
+   * The {@link Generator} instance that performed the semantic analysis
+   * creating this {@link ModuleNode} instance.
+   */
+  public final Generator SemanticChecker;
```
Nit: `semanticChecker`

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCSourceBreakpoint.java:
```diff
@@ -55,8 +68,95 @@ public TLCSourceBreakpoint(final String module, final SourceBreakpoint s) {
+		// in module to make syntax parser happy, then throw away the
+		// encapsulating module. Future syntax parser updates might allow
+		// parsing the operator definition directly.
+		final String wrapper = "---- MODULE Ignore ----\n%s == %s\n====";
```
Perhaps, e.g. `Ignore21384ylshfutfaoug2` because `Ignore` could also be a user spec.

ahelwer replied:
The Ignore name will never make it to semantic checker so it won't be checked for conflicts; the syntax parser doesn't check for name collisions.

lemmy replied:
Please add a corresponding note to the comment.

ahelwer commented:
Also thought of a tricky problem. Implicitly, the new operator is parsed at the very bottom of the file; however, consider this spec:
```tla
---- MODULE Test ----
op(x) == (* want to debug here *) TRUE
VARIABLE x
====
```
We would generate an operator with definition `__BreakpointExpr__0(x) == ...` but this would then cause a semantic failure since `x` is already defined in the global scope by the end of the file. Something to think about. It might be that in addition to generating a unique `__BreakpointExpr__n` name for the operator, we also need to generate unique names for the parameters, ending up with something like `__BreakpointExpr__0(__bpParam__x) == ...` then when resolving the breakpoint we extract "`x`" from `__bp_Param__x` and plug that name into the stack frame.

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCSourceBreakpoint.java:
```diff
@@ -55,8 +68,95 @@ public TLCSourceBreakpoint(final String module, final SourceBreakpoint s) {
+		// action-level or below.
+		Errors levelCheckingErrors = new Errors();
+		SemanticNode.setError(levelCheckingErrors); // Annoyingly static
+		if (!semanticOp.levelCheck(1) || levelCheckingErrors.isFailure()) {
```
Consider adding a zero-arity `LevelNode#levelCheck` that invokes `LevelNode#levelCheck` with `LevelNode#levelChecked + 1`

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCSourceBreakpoint.java:
```diff
@@ -55,8 +69,105 @@ public TLCSourceBreakpoint(final String module, final SourceBreakpoint s) {
+			ToolIO.err.println(levelCheckingErrors.toString());
+			return null;
+		}
+		if (LevelConstants.ConstantLevel == semanticOp.level
```
`tla2sany.semantic.LevelNode.getLevel()`

More importantly, can you make the context of the breakpoint determine the maximum level of the debug expression? For instance, the maximum level of a breakpoint expression for a constant formula is constant, and so on.

Related:
```java
tla2sany.semantic.SemanticNode.pathTo(Location)
```

lemmy commented on tlatools/org.lamport.tlatools/src/tla2sany/semantic/ModuleNode.java:
```diff
@@ -419,6 +426,20 @@ public final OpDefNode[] getOpDefs() {
+   * definition can be surfaced when the cache is rebuilt; that is what this
+   * function does.
+   */
+  public void clearOpDefCache() {
```
Are you certain that the underlying `ModuleNode#ctxt` cannot be mutate after `getOpDefs` is invoked the first time by legacy code?

ahelwer replied:
I'm not sure; this is something to explore with this prototype. The role of the context class is still a bit mysterious to me and I don't know how it connects to other things in the parser. It is home to a piece of static global state I want to get rid of though, so understanding it is hopefully in my future.

ahelwer commented:
@lemmy all right I'm at the end of my expertise here, I have a test setup and have debugged it all the way to where it constructs the context to evaluate the breakpoint expression. Unfortunately I get an empty context no matter what, it doesn't even contain the global variable and constant values. For now I'm just testing expressions that depend only on global identifiers to see whether this works at all before investing time on doing the whole local identifier analysis thing.

lemmy commented:
> @lemmy all right I'm at the end of my expertise here, I have a test setup and have debugged it all the way to where it constructs the context to evaluate the breakpoint expression. Unfortunately I get an empty context no matter what, it doesn't even contain the global variable and constant values. For now I'm just testing expressions that depend only on global identifiers to see whether this works at all before investing time on doing the whole local identifier analysis thing.

Are you saying that `tlc2.debug.TLCStackFrame#ctxt` is always empty?  Did you try with *inline* breakpoints?

Related: 4a2586e2b24afdf633cf3c28c4900449f9793656

ahelwer commented:
Well in my test I have this module:
```tla
---- MODULE ExpressionBreakpointTest ----
EXTENDS Naturals
VARIABLE i
CONSTANT j
op(k) ==
  \A l \in 0 .. k :
    LET m == i IN
      (* breakpoint 1 here *) TRUE
(* breakpoint 2 here *) Inv == op(i)
Init == i = 0
Next == IF i < j THEN i' = i + 1 ELSE UNCHANGED i
====
---- CONFIG ExpressionBreakpointTest ----
CONSTANT j = 10
INVARIANT Inv
INIT Init
NEXT Next
====
```
and yeah when I run the debugger and it hits breakpoint 1 the context I get from `getContext()` is empty.

ahelwer commented:
Okay so the Context seems to only include local identifier values, not global identifier values. How do I access global identifier values?

ahelwer commented:
The evaluator access state-level variables through the state; the missing part now is the value of constants. Are constant values inlined in TLA+ expressions? If so will need to perform that on the user-supplied breakpoint expression.

lemmy commented:
> The evaluator access state-level variables through the state; the missing part now is the value of constants. Are constant values inlined in TLA+ expressions? If so will need to perform that on the user-supplied breakpoint expression.

There was no need to explicitly handle variables and constants with the “pre-startup” or “hardcoded” breakpoint expressions because `tlc2.tool.impl.SpecProcessor` automatically handled them. However, `tlc2.tool.impl.SymbolNodeValueLookupProvider` should be able to handle most, if not all, of the job. Alternatively, `tlc2.tool.impl.SpecProcessor.getConstantDefs()` could be relevant.

ahelwer commented:
There's a missing link somewhere unfortunately. Even just trying to evaluate the breakpoint expression `4 > 3` results in the following error:
>tlc2.tool.EvalException: Attempted to apply the operator overridden by the Java method public static tlc2.value.impl.BoolValue tlc2.module.Integers.GT(tlc2.value.impl.Value,tlc2.value.impl.Value), but it produced the following error:
Cannot invoke "tlc2.value.impl.Value.toString()" because "x" is null

So it seems just running it through the semantic analysis engine isn't enough, there is some additional expression processing that needs to happen before it can be evaluated.

lemmy commented:
You will have to process the `Ignore` module with `SpecProcessor#processConstants`.

https://github.com/tlaplus/tlaplus/blob/68b6f50f7e36b19dae03007620df928cb4d1b419/tlatools/org.lamport.tlatools/src/tlc2/tool/impl/SpecProcessor.java#L1769-L1834

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCSourceBreakpoint.java:
```diff
@@ -55,8 +58,9 @@ public TLCSourceBreakpoint(final String module, final SourceBreakpoint s) {
 			}
 		}
 		hits = h;
+		this.condition = TLCBreakpointExpression.process(semanticRoot, s.getCondition());
```
This needs to check if s.getCondition is an existing definition such as `TypeOK`, ... that doesn't need processing.

ahelwer replied:
If the user puts in an existing definition like `TypeOK` I think the processing code will ultimately result in what the user intends; it could be short-circuited by looking up existing definitions in the module but this sort of special casing would also add another code path to test.

lemmy replied:
How does it ultimately result in what the user intends?  Does it create a module containing a single definition, `TypeOK_139488708`, which delegates to `TypeOK` in the original specification, similar to how it would handle true ad-hoc/on-the-fly/dynamic breakpoint expressions?

ahelwer replied:
It would create something like
```tla
---- MODULE __BreakpointModule__12345678
EXTENDS OriginalModuleName
__BreakpointDef__12345678(i, j, k) == TypeOK
====
```
and then when the breakpoint is evaluated it should just evaluated `TypeOK`.

lemmy commented:
> You will have to process the `Ignore` module (or the `semanticOp`) with `SpecProcessor#processConstants`.

Processing the breakpoint expression within the scope of the `semanticRoot` is likely the most suitable approach. However, the `SpecProcessor` is unable to incrementally process the `semanticRoot`, and enhancing it would be highly error-prone (a debugger's design guideline is not to modify the existing interpreter). Therefore, I propose attempting to make the `Ignore` module an *inner* module of the `semanticRoot`.

lemmy commented:
More debugging indicates that the inner module approach may not be sufficient (the inner module doesn’t appear to be encapsulated within its own `ModuleNode`). If confirmed, `Ignore` could also extend `semanticRoot`.

ahelwer commented:
>Processing the breakpoint expression within the scope of the semanticRoot is likely the most suitable approach. However, the SpecProcessor is unable to incrementally process the semanticRoot, and enhancing it would be highly error-prone

We could punt on this work until after I'm finished with the semantic test corpus work, which would give us more confidence trying to modify it to accept incremental processes of this type.

ahelwer commented:
>More debugging indicates that the inner module approach may not be sufficient (the inner module doesn’t appear to be encapsulated within its own ModuleNode). If confirmed, Ignore could also extend semanticRoot.

That could be another possible approach. It should be possible to find & iterate through the `ExternalModuleTable` to ensure generation of a module name that does not conflict with any others.

ahelwer commented:
>More debugging indicates that the inner module approach may not be sufficient (the inner module doesn’t appear to be encapsulated within its own ModuleNode). If confirmed, Ignore could also extend semanticRoot.

I've now implemented this approach. However, I don't know how to easily construct a `SpecProcessor` instance with the state available while defining a new breakpoint; do you know what would need to be pulled in for that?

lemmy commented:
Why don't you reuse the existing `tool.getSpecProcessor` instance?

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java:
```diff
@@ -0,0 +1,161 @@
+		long suffix = 0L;
+		do {
+			unusedName = String.format(pattern, Long.toString(suffix));
+			suffix++;
```
Minor: The likelihood of a collision is significantly reduced if the suffix is `System.currentTimeMillis()`, which aligns with the approach employed by the Toolbox.

lemmy commented:
The TLA+ Debugger also supports "Watch" expressions.  AFAICT, Watch and Breakpoint expressions will be able to share most of the implementation.

https://www.youtube.com/watch?v=r7kGbjcSaKU

lemmy commented:
The branch `mku-pr1099` is based on this pull request and modifies the `SpecProcessor` to process breakpoint expressions correctly. Specifically, `SpecProcessor` now takes another snapshot of the `SpecProcessor#defns` before module overrides, constants definitions, ... are processed. This modification allows context-free breakpoint expressions, such as those found in the `EchoDebuggerTest` module, to successfully evaluate.

ahelwer commented:
Impressive! it can now process expressions like:
- `4 > 3` (built-in constants)
- `j > 3` (user-defined constants)
- `(i + 3) > j` (user-defined constants and variables)

So looks like it's time to identify local-scope identifiers and add them to the operator definition.

ahelwer commented:
It's proving a bit difficult to calculate what local variables can be accessed from any given expression node. Some of the semantic tree nodes have a `SymbolTable` instance associated with them but it isn't a copy, so as the semantic checker process continued the table got modified from under it. So unless some magic is being done in the spec processor that can surface this information I think I'll have to modify the semantic checking process to record some additional info in each semantic node.

lemmy commented:
Assuming `p1` and `p2` are local identifiers within the scope of the breakpoint expression, would it make things easier if the user to be required to write something like `LET p1 = TRUE, p2 = TRUE IN variable = p1 + p2`?  In other words, let the user identify the local identifiers for you.

ahelwer commented:
For now perhaps I will do a best-effort crawl back up the path from breakpoint node to module node, recording identifiers introduced by:
1. Formal parameter nodes of operators like `f(a,b,c) == `
2. Identifiers introduced in `LET/IN` blocks
3. Identifiers introduced in quantifiers

This will probably miss some cases (lambdas? idk) but will at least be a start.

ahelwer commented:
Local identifier resolution is now implemented and seems to basically work, which means we can transition this out of a draft PR!

lemmy commented on tlatools/org.lamport.tlatools/src/tla2sany/semantic/SymbolTable.java:
```diff
@@ -2,7 +2,9 @@
 package tla2sany.semantic;
 
-import tla2sany.utilities.Stack;
+import java.util.ArrayDeque;
```
Why is this change necessary? The collections in `java.util` and those in `tla2sany.util` are unlikely to be fully compatible.

lemmy replied:
If this is absolutely necessary and cannot be avoided, isolate the code into a dedicated commit.

ahelwer replied:
I will cherry-pick this commit out into another branch for submission in the future. Primarily I made it because the state of our custom-written `Stack` class is hard to understand when viewed in the debugger. It would be nice to get rid of it eventually.

lemmy commented on tlatools/org.lamport.tlatools/src/tla2sany/semantic/SemanticNode.java:
```diff
@@ -250,7 +250,7 @@ public LinkedList<SemanticNode> get() {
 					@Override
 					public void preVisit(final SemanticNode node) {
-						if (location.equals(node.getLocation())) {
+						if (location.equals(node.getLocation()) || (0 == node.getChildren().length && node.getLocation().includes(location))) {
```
`SemanticNode#getChildren` can return null > possible NPE.

lemmy replied:
Not just hypothetical; just happened when toying with the PR.

lemmy replied:
Note the comment about elems of getChildren potentially being `null` in `tla2sany.semantic.SemanticNode#getListOfChildren`

lemmy replied:
https://github.com/tlaplus/tlaplus/commit/81ed3ffbbcb71ef70c3a1f6721c9a113a77f09c1

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java:
```diff
@@ -0,0 +1,222 @@
+		final String wrapper = "---- MODULE %s ----\nEXTENDS %s\n%s == %s\n====";
+		String wrappedConditionExpr = String.format(wrapper, bpModName, rootModName, bpOpDef, conditionExpr);
+		byte[] wrappedConditionExprBytes = wrappedConditionExpr.getBytes(StandardCharsets.UTF_8);
+		InputStream sourceCode = new ByteArrayInputStream(wrappedConditionExprBytes);
```
Should this stream be closed explicitly? Perhaps, be wrapped in a “try-with-resources” block?

ahelwer replied:
Per Java docs closing a `ByteArrayInputStream` has no effect: https://docs.oracle.com/en/java/javase/11/docs/api/java.base/java/io/ByteArrayInputStream.html#close()

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCSourceBreakpoint.java:
```diff
@@ -27,20 +27,26 @@
 		final int column = getColumn() != null ? getColumn() : 1;
 		//TODO: If the location spans lines, getLine() + 1 should be the endline (second line parameter of Location).
-		location = new Location(module, getLine() + 1, column, getLine(), column + 1);
+		location = new Location(module, getLine(), column, getLine(), column + 1);
```
This change is likely the cause of the test failures. I believe there was a mismatch between the SANY and the DAP locations.

ahelwer replied:
I recall I ran into trouble getting a match using the `SemanticNode.pathTo` method with the `+ 1` added; will investigate further.

lemmy replied:
Related, breakpoint validation/verification (hollow breakpoint vs. Red breakpoint) no longer seems to works.  Bisecting should be the fastest to find the offending commit

ahelwer replied:
Like bisect within these changes or in the repo in general?

ahelwer replied:
It seems like the `+ 1` works with my changes to how `pathTo` works (accept if node is a leaf node and the location is contained within it).

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java:
```diff
@@ -0,0 +1,222 @@
+package tlc2.debug;
```
Copyright header

lemmy commented on tlatools/org.lamport.tlatools/src/tla2sany/semantic/OpDefNode.java:
```diff
@@ -1480,6 +1480,11 @@ protected String getNodeRef() {
     }
   }
+  
+  public String toString() {
```
This appears to be a debugging tool that might no longer be relevant.

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/tool/impl/SpecProcessor.java:
```diff
@@ -123,7 +123,8 @@ public class SpecProcessor implements ValueConstants, ToolGlobals {
     private Set<OpDefNode> processedDefs;
     private SpecObj specObj;
-    private Defns snapshot;
+    private Defns preConstantSnapshot;
```
The changes made to the `SpecProcessor` will also need to be isolated into a separate commit. This is my TODO.

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java:
```diff
@@ -0,0 +1,222 @@
+			ToolIO.err.println("ERROR: unable to find breakpoint expression op " + bpOpName);
+		}
+
+		if (!(LevelConstants.ConstantLevel == bpOp.level || LevelConstants.VariableLevel == bpOp.level
```
This level check is sufficient for evaluating *Watch* expressions (not breakpoints). A Watch expression is evaluated for the currently active stack frame, regardless of whether that frame matches the expression’s level. If the evaluation fails, the debugger displays an appropriate message indicating that the Watch expression failed to evaluate.

ahelwer replied:
Does that mean I shouldn't implement the
>Conditionally restrict operator action level depending on breakpoint location

feature?

lemmy replied:
Definitely not for Watch expressions, but I could imagine that novice users would appreciate it if the debugger would notify them if a breakpoint expression's level exceeds the scope's level. That said, does DAP even provide a mechanism to notify the user beyond visual breakpoint validation/verification (hollow vs. red breakpoint)?

ahelwer replied:
Not that I was able to find, other than hollow breakpoints + debug log.

lemmy commented:
> The TLA+ Debugger also supports "Watch" expressions. AFAICT, Watch and Breakpoint expressions will be able to share most of the implementation.
> 
> https://www.youtube.com/watch?v=r7kGbjcSaKU

Ad-hoc/on-the-fly/dynamic Watch expressions have been implemented in branch `mku-pr1099`, which has cherry-picked the commits from this branch.

lemmy commented:
> Local identifier resolution is now implemented and seems to basically work, which means we can transition this out of a draft PR!

Is the following case (`bciFastPath` let expression) supported?  Hovering (right) resolves `bciFastPath`.

<img width="1263" alt="Screenshot 2024-12-19 at 6 48 07 PM" src="https://github.com/user-attachments/assets/4fdafe47-477f-4a37-82ac-60ed21082a51" />

ahelwer commented:
I added a bunch of tests for the `getScopedIdentifier` method. It supports the `bciFastPath` case, and I also added support for `LAMDA` expressions (no idea whether the debugger itself supports breaking inside a `LAMBDA` but whatever).

ahelwer commented:
There's a failure in `EWD998ChanDebuggerTest` and I'm not really sure how to debug it:

```
org.junit.ComparisonFailure: expected:<[OpAppl]Node> but was:<[At]Node>
	at org.junit.Assert.assertEquals(Assert.java:117)
	at org.junit.Assert.assertEquals(Assert.java:146)
	at tlc2.debug.EWD998ChanDebuggerTest.testSpec(EWD998ChanDebuggerTest.java:374)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at util.IsolatedTestCaseRunner.run(IsolatedTestCaseRunner.java:71)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:757)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)

java.lang.AssertionError: expected:<0> but was:<-1>
	at org.junit.Assert.fail(Assert.java:89)
	at org.junit.Assert.failNotEquals(Assert.java:835)
	at org.junit.Assert.assertEquals(Assert.java:647)
	at org.junit.Assert.assertEquals(Assert.java:633)
	at tlc2.tool.liveness.ModelCheckerTestCase.assertExitStatus(ModelCheckerTestCase.java:267)
	at tlc2.tool.liveness.ModelCheckerTestCase.tearDown(ModelCheckerTestCase.java:263)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.RunAfters.invokeMethod(RunAfters.java:46)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:33)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at util.IsolatedTestCaseRunner.run(IsolatedTestCaseRunner.java:71)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:757)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
```

lemmy commented:
> There's a failure in `EWD998ChanDebuggerTest` and I'm not really sure how to debug it:
> 
> ```
> org.junit.ComparisonFailure: expected:<[OpAppl]Node> but was:<[At]Node>
> 	at org.junit.Assert.assertEquals(Assert.java:117)
> 	at org.junit.Assert.assertEquals(Assert.java:146)
> 	at tlc2.debug.EWD998ChanDebuggerTest.testSpec(EWD998ChanDebuggerTest.java:374)
> 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
> 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
> 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
> 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
> 	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
> 	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
> 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
> 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
> 	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
> 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
> 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
> 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
> 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
> 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
> 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
> 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
> 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
> 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
> 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
> 	at util.IsolatedTestCaseRunner.run(IsolatedTestCaseRunner.java:71)
> 	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
> 	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
> 	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
> 	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:757)
> 	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
> 	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
> 
> java.lang.AssertionError: expected:<0> but was:<-1>
> 	at org.junit.Assert.fail(Assert.java:89)
> 	at org.junit.Assert.failNotEquals(Assert.java:835)
> 	at org.junit.Assert.assertEquals(Assert.java:647)
> 	at org.junit.Assert.assertEquals(Assert.java:633)
> 	at tlc2.tool.liveness.ModelCheckerTestCase.assertExitStatus(ModelCheckerTestCase.java:267)
> 	at tlc2.tool.liveness.ModelCheckerTestCase.tearDown(ModelCheckerTestCase.java:263)
> 	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
> 	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
> 	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
> 	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
> 	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
> 	at org.junit.internal.runners.statements.RunAfters.invokeMethod(RunAfters.java:46)
> 	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:33)
> 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
> 	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
> 	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
> 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
> 	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
> 	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
> 	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
> 	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
> 	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
> 	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
> 	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
> 	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
> 	at util.IsolatedTestCaseRunner.run(IsolatedTestCaseRunner.java:71)
> 	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:93)
> 	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:40)
> 	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:529)
> 	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:757)
> 	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:452)
> 	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:210)
> ```

It is due to the following change:

https://github.com/tlaplus/tlaplus/pull/1099/files#diff-7ecdbd526fe811798f77baebd73fd5cf826805f74199fbd832a3037cb505bda7R256-R257

ahelwer commented:
Okay I fixed it, UTs should pass now. The only remaining thing is to fix up my `ExpressionBreakpointTest` to actually assert some things.

lemmy commented:
I force-pushed [two commits](https://github.com/tlaplus/tlaplus/compare/master...mku-pr1099) to branch [mku-pr1099](https://github.com/tlaplus/tlaplus/compare/master...mku-pr1099) that (logically) replace the `SpecProcessor`-related changes in commit https://github.com/tlaplus/tlaplus/pull/1099/commits/3287902cfbf52774d2d194d25542cd69f9c532d5

lemmy commented:
Below is the prototype to enable support for dynamic Watch expressions:

```diff
commit 3de9dc64aafcc4f70439046dbd7ac82aa90137de
Author: Markus Alexander Kuppe <github.com@lemmster.de>
Date:   Thu Dec 19 18:36:40 2024 -0800

    Initial support for ad-hoc/on-the-fly/dynamic Watch expressions.

    [Feature][TLC][Debuggeer]

    Signed-off-by: Markus Alexander Kuppe <github.com@lemmster.de>

diff --git a/tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java b/tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java
index 10a99000e..4e64b9499 100644
--- a/tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java
+++ b/tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java
@@ -38,12 +38,19 @@ public class TLCBreakpointExpression {
 	 * @param conditionExpr The unparsed breakpoint expression.
 	 * @return A breakpoint expression, or null if parsing failed.
 	 */
-	public static OpDefNode process(
-			final SpecProcessor processor,
-			final ModuleNode semanticRoot,
-			final Location location,
-			final String conditionExpr
-	) {
+	public static OpDefNode process(final SpecProcessor processor, final ModuleNode semanticRoot,
+			final Location location, final String conditionExpr) {
+		final List<String> paramNames = getScopedIdentifiers(semanticRoot, location);
+		return process(processor, semanticRoot, paramNames, conditionExpr);
+	}
+
+	public static OpDefNode process(final SpecProcessor processor, final ModuleNode semanticRoot,
+			final String conditionExpr) {
+		return process(processor, semanticRoot, new ArrayList<>(), conditionExpr);
+	}
+
+	public static OpDefNode process(final SpecProcessor processor, final ModuleNode semanticRoot,
+			final List<String> paramNames, final String conditionExpr) {
 		if (null == conditionExpr || conditionExpr.isBlank()) {
 			return null;
 		}
@@ -53,7 +60,6 @@ public class TLCBreakpointExpression {
 		final String bpModName = generateUnusedName(semanticRoot, "__BreakpointModule__%s");
 		ToolIO.out.println("BPExpr: wrapping with module \"" + bpModName + "\"");
 		final String bpOpName = generateUnusedName(semanticRoot, "__BreakpointExpr__%s");
-		final List<String> paramNames = getScopedIdentifiers(semanticRoot, location);
 		final String params = paramNames.size() > 0 ? "(" + String.join(", ", paramNames) + ")" : "";
 		final String bpOpDef = bpOpName + params;
 		ToolIO.out.println("BPExpr: wrapping with op \"" + bpOpDef + "\"");
diff --git a/tlatools/org.lamport.tlatools/src/tlc2/debug/TLCStackFrame.java b/tlatools/org.lamport.tlatools/src/tlc2/debug/TLCStackFrame.java
index d48ea4a04..09e71cad9 100644
--- a/tlatools/org.lamport.tlatools/src/tlc2/debug/TLCStackFrame.java
+++ b/tlatools/org.lamport.tlatools/src/tlc2/debug/TLCStackFrame.java
@@ -545,8 +545,17 @@ public class TLCStackFrame extends StackFrame {
 		if (name == null) {
 			return new EvaluateResponse();
 		}
+
+		// 1) Try to resolve the name to an existing expression.
 		final ModuleNode module = tool.getSpecProcessor().getRootModule();
-		return getWatch(module.getOpDef(name));
+
+		// 2) Expression doesn't exist. Check if name is an ad-hoc expression.
+		OpDefNode odn = module.getOpDef(name);
+		if (odn == null) {
+			odn = TLCBreakpointExpression.process(tool.getSpecProcessor(), module, name);
+		}
+
+		return getWatch(odn);
 	}

 	public EvaluateResponse getWatch(final OpDefNode odn) {
```

ahelwer commented:
Do we want to merge this PR and then work on watch expressions on top of it as an additional feature?

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java:
```diff
@@ -0,0 +1,269 @@
+
+		final String rootModName = semanticRoot.getName().toString();
+		final String bpModName = generateUnusedName(semanticRoot, "__BreakpointModule__%s");
+		ToolIO.out.println("BPExpr: wrapping with module \"" + bpModName + "\"");
```
The logging here and below seems a bit excessive.

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java:
```diff
@@ -0,0 +1,269 @@
+/*******************************************************************************
+ * Copyright (c) 2023 Linux Foundation. All rights reserved.
```
Class was created in 2024.

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java:
```diff
@@ -0,0 +1,269 @@
+
+		Errors semanticLog = new Errors();
+		SemanticNode.setError(semanticLog); // Annoyingly static
+		Generator semanticChecker = new Generator(semanticRoot.semanticChecker.moduleTable, semanticLog);
```
There is no need to change `ModuleNode` (see above) because this can use `SpecProcessor#getModuleTbl`.

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java:
```diff
@@ -0,0 +1,269 @@
+	 * @param location     The breakpoint location.
+	 * @return A list of local identifiers accessible at the given location.
+	 */
+	static Set<String> getScopedIdentifiers(
```
Should/Could this be a method of `ModuleNode`?

ahelwer replied:
I think it's a bit half-baked to move into ModuleNode. I would rather fix this by refactoring each node to keep a snapshot of the `tla2sany.semantic.SymbolTable` stack around. Not something that can be done without the semantic test corpus.

lemmy replied:
Out of scope for this PR.

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java:
```diff
@@ -0,0 +1,269 @@
+	 * @param pattern A base pattern in which a number can be interpolated.
+	 * @return A name that is unique within the context of the module.
+	 */
+	private static String generateUnusedName(ModuleNode semanticRoot, String pattern) {
```
Again, method of `ModuleNode`?

ahelwer replied:
Seems like a good candidate yeah

lemmy commented on tlatools/org.lamport.tlatools/test/util/ParserInterface.java:
```diff
@@ -0,0 +1,108 @@
+   */
+  public static ModuleNode parse(String input) {
+    SyntaxTreeNode syntaxRoot = processSyntax(input);
+    if (null == syntaxRoot) return null;
```
Please add curly braces.

lemmy commented on tlatools/org.lamport.tlatools/test/util/ParserInterface.java:
```diff
@@ -0,0 +1,108 @@
+ * tests. This class can be viewed as a prototype of a future API to be
+ * integrated into SANY itself.
+ */
+public class ParserInterface {
```
Terminology: "Interface" yet it is a Java class.

ahelwer replied:
I was thinking of calling it API. Do you want to use ParserAPI or ParserApi casing?

lemmy commented on tlatools/org.lamport.tlatools/test/tlc2/debug/GetScopedIdentifiersTests.java:
```diff
@@ -0,0 +1,95 @@
+   * Run all the test cases and check the identifiers are found as expected.
+   */
+  @Test
+  public void test() {
```
Let's stick to ordinary junit tests and one @Test method per expression so that we know which one fails (if any).

ahelwer replied:
This is a deficiency in JUnit 4 (which we use) that was fixed in JUnit 5 with the addition of [dynamic tests](https://junit.org/junit5/docs/current/user-guide/#writing-tests-dynamic-tests), so each of the test cases here could be generated as a separate test at runtime. Testing has broadly moved away from hardcoded tests toward tests-as-data or even auto-generated test cases. I find it much easier to add another line in the array compared to adding another test method.

Calvin-L replied:
JUnit 4 supports [parameterized tests](https://github.com/junit-team/junit4/wiki/Parameterized-tests), which is the best of both worlds:
 - You can have a simple array of test inputs
 - Each input is treated as a separate test in JUnit's reports

lemmy replied:
+1 for parameterized tests, which have already been used in e.g. https://github.com/tlaplus/tlaplus/blob/master/tlatools/org.lamport.tlatools/test/tlc2/value/impl/SubsetEnumeratorTest.java

lemmy commented on tlatools/org.lamport.tlatools/test/util/ParserInterface.java:
```diff
@@ -0,0 +1,108 @@
+package util;
```
This class seems largely redundant to TLCBreakpointExpression?  Can this class be moved out of the test package and be used by TLCBreakpointExpression?

ahelwer replied:
I agree it is redundant; my sort of long-term goal is to develop this API class according to the needs of the test code, and then eventually introduce it in a `tla2sany.api` java package once it has been sufficiently refined. Ideally it would then not change very much. However a few things remain before I think it's fully baked:
1. Remove `SemanticNode.error` and `Context.initialContext` global static state
2. Figure out a good module resolution process
3. Figure out a good incremental parse API, allowing users to add modules & definitions and hopefully ghost expressions to be evaluated at an arbitrary point in the tree.
4. Figure out how the API would integrate with a systematic error reporting system as sketched in https://github.com/tlaplus/rfcs/issues/15

lemmy replied:
Let’s consider this out of scope for this PR.

lemmy commented on tlatools/org.lamport.tlatools/test/tlc2/debug/ExpressionBreakpointTest.java:
```diff
@@ -0,0 +1,66 @@
+		if (result instanceof IntValue) {
+			return ((IntValue)result).val;
+		} else if (result instanceof LazyValue) {
+			// Evade cache guard by guessing value from small range.
```
Why is this necessary?

ahelwer replied:
I don't know how to access the value of lazy things. It seems like it is guarded fairly well.

lemmy replied:
```java
return ((IntValue) lazy.getValue(frame.tool, frame.getS(), frame.getT(), EvalControl.Clear)).val;
```

lemmy replied:
Even shorter:

```java
return ((IntValue) frame.unlazy(lazy)).val;
```

lemmy commented on tlatools/org.lamport.tlatools/src/tla2sany/semantic/SemanticNode.java:
```diff
@@ -228,14 +231,20 @@ public T get() {
+	 *         matches the given location constraints.
 	 */
-	public LinkedList<SemanticNode> pathTo(final Location location) {
+	public LinkedList<SemanticNode> pathTo(final Location location, boolean requireExactLocationMatch) {
```
The 1-arity `pathTo` could delegate to the new, 2-arity pathTo.

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCStackFrame.java:
```diff
@@ -499,7 +499,7 @@ public EvaluateResponse get(final EvaluateArguments ea) {
 
 			final Location location = Location.parseCoordinates(moduleName, u.getFragment());
-			final LinkedList<SemanticNode> path = tool.getModule(location.source()).pathTo(location);
+			final LinkedList<SemanticNode> path = tool.getModule(location.source()).pathTo(location, true);
```
https://github.com/tlaplus/tlaplus/pull/1099/files#r1899790891

lemmy commented on tlatools/org.lamport.tlatools/test/tla2sany/semantic/TestLevelChecking.java:
```diff
@@ -1,14 +1,10 @@
 package tla2sany.semantic;
```
Copyright header

lemmy commented on tlatools/org.lamport.tlatools/test/tlc2/debug/ExpressionBreakpointTest.java:
```diff
@@ -0,0 +1,66 @@
+package tlc2.debug;
```
Copyright header

lemmy commented on tlatools/org.lamport.tlatools/test/tlc2/debug/GetScopedIdentifiersTests.java:
```diff
@@ -0,0 +1,104 @@
+package tlc2.debug;
```
Copyright header

ahelwer replied:
Do we really need the MIT header in every file?

lemmy replied:
IANAL, but this is standard practice, and most IDEs typically fold the license header automatically.

lemmy commented on tlatools/org.lamport.tlatools/test/util/ParserAPI.java:
```diff
@@ -0,0 +1,110 @@
+package util;
```
Copyright header

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCSourceBreakpoint.java:
```diff
@@ -27,14 +27,20 @@
 	private final int hits;
 	private final Location location;
+	// Use Getter
+	public final OpDefNode condition;
```
`condition` can be private if the implementation of the method `TLCStateStackFrame#matchesExpression` is moved into `TLCSourceBreakpoint` (`TLCStateStackFrame.matchesExpression` then delegates to `TLCSourceBreakpoint#matchesExpression).  Moreover, the impl no longer has to check `bp.getCondition() != null && !bp.getCondition().isEmpty()`.

ahelwer replied:
I looked into doing this but there is a fair bit of `TLCStateStackFrame`-specific stuff in the `matchesExpression` function:
- The context
- The tool
- `getS()`
- `getT()`

lemmy commented on tlatools/org.lamport.tlatools/test/tlc2/debug/ExpressionBreakpointTest.java:
```diff
@@ -0,0 +1,66 @@
+		// Break at init
+		debugger.continue_();
+		
+		// Break at bp
```
This block can be simplified into:

```java
// Break at bp
final TLCStateStackFrame current = (TLCStateStackFrame) debugger.continue_()[0];
assertEquals(2, ((IntValue) current.getT().getVals().get(UniqueString.of("i"))).val);
final Map<String, String> allVariables = new HashMap<>();
allVariables.put("k", "5");
allVariables.put("l", "4");
assertTLCStateFrame(current, 8, 8, RM, allVariables);
```

`getStateVal` and `getContextVal` become obsolete.

ahelwer replied:
I kind of like the test as it is. It's a bit more clear that a property is being tested instead of hardcoding specific expected values.

lemmy replied:
For consistency, please stick to the style that is used by the other `tlc2.debug` tests.

ahelwer replied:
I will change it, but I would like to lodge feedback that I find these type of tests confusing since it is never clear if they're just recording & playing back behavior or actually testing a property that matters. A good example is the test failure that was caused by changing the `pathTo` function, which manifested in a very weird way.

lemmy commented on tlatools/org.lamport.tlatools/src/tlc2/debug/TLCBreakpointExpression.java:
```diff
@@ -0,0 +1,245 @@
+		byte[] wrappedConditionExprBytes = wrappedConditionExpr.getBytes(StandardCharsets.UTF_8);
+		InputStream sourceCode = new ByteArrayInputStream(wrappedConditionExprBytes);
+		TLAplusParser parser = new TLAplusParser(sourceCode, StandardCharsets.UTF_8.name());
+		boolean syntaxParseSuccess = parser.parse();
```
A `conditionExpr` such as "Does not exist" cause `parse()` to print:

```
***Parse Error***
Was expecting "==== or more Module body"
Encountered "not" at line 3, column 39 and token "Does" 
```

lemmy replied:
See `EWD998ChanDebuggerTest` after cherry-picking https://github.com/tlaplus/tlaplus/commit/8baf0d86ac3d14f26c11c3b3261a6b8f2e70eb54

lemmy replied:
To alleviate the issue, the error message could be modified to include the module name, thereby prompting users to recognize that the problem lies within the debug module.

ahelwer replied:
Added some additional error logging

lemmy commented:
> Do we want to merge this PR and then work on watch expressions on top of it as an additional feature?

The size of the related commit (https://github.com/tlaplus/tlaplus/commit/8baf0d86ac3d14f26c11c3b3261a6b8f2e70eb54) doesn't warrant a dedicated PR.  Please just cherry-pick on top of your branch.

lemmy reviewed (approved):
LGTM once my unresolved comments have been addressed.

lemmy commented on tlatools/org.lamport.tlatools/test/util/ParserAPI.java:
```diff
@@ -0,0 +1,133 @@
+      ModuleNode semanticRoot = gen.generate(parseTree);
+      return log.isSuccess() ? semanticRoot : null;
+    } catch (AbortException e) {
+      Assert.fail(e.toString() + log.toString());
```
Why does this catch the specific `AbortException` to replace it with the generic `TLCRuntimeException`?

ahelwer replied:
Temporary. The error handling of this API is not yet figured out. Probably it will eventually be much more elaborate as https://github.com/tlaplus/rfcs/issues/15 is worked on.

lemmy replied:
Please use TODO markers for incomplete work.

lemmy commented on tlatools/org.lamport.tlatools/test/util/ParserAPI.java:
```diff
@@ -0,0 +1,133 @@
+      return log.isSuccess() ? semanticRoot : null;
+    } catch (AbortException e) {
+      Assert.fail(e.toString() + log.toString());
+      return null;
```
Dead code.

ahelwer replied:
Yes but the Java compiler is not smart enough to know that and raises a compilation error that the method needs to return a value in all paths.

lemmy commented on tlatools/org.lamport.tlatools/test/tlc2/debug/GetScopedIdentifiersTests.java:
```diff
@@ -0,0 +1,134 @@
+  public static TestCase[] testCases() {
+    return new TestCase[] {
+      new TestCase("op ≜", ""),
+      new TestCase("op(i, j, k) ≜", "", "j", "i", "k"),
```
This doesn't have to be changed, but in general, I want to use ASCII in our test suite to maintain compatibility with older versions of tla2tools that were released before the Unicode feature was introduced. This ensures we can run new tests on those older versions should this become necessary.

lemmy commented:
Thanks, logically merged into `master`

lemmy commented:
`tlc2.debug.TLCDebuggerExpression.process(SpecProcessor, ModuleNode, String)` doesn't handle expressions of the form `LET M == INSTANCE SomeModule IN M!SomeOp` if `SomeModule` appears nowhere else.  The fact that the `ExternalModuleTable` contains no entry for `SomeModule` makes  `tla2sany.semantic.Generator.processModuleDefinition(TreeNode, Vector, Vector, ModuleNode)` add a parse error.

ahelwer commented:
I can see why that would be desirable to write. I will think about how to handle it.
